test/testtokenize.cpp

```diff
@@ -7904,6 +7966,25 @@ class TestTokenizer : public TestFixture {
         ASSERT_EQUALS("b(", testAst("class a { void b() & {} };"));
     }
 
+    void astnewscoped() {
+        auto generateAst = [&](std::string code) {
+            Settings settings;
+            Library library;
+            auto tokenizer = std::make_unique<Tokenizer>(&settings, this);
+            std::istringstream istr(code);
+            tokenizer->tokenize(istr, "test.cpp");
+            return tokenizer;
+        };
+
+        //tokenizer.
+        bool match1 = SimpleAst{"return", {"new", {"A"}}}.matches(*generateAst("return new A;"));
```

Comment:
About `SimpleAst`.. I am not sure about it. Right now it seems the test code would become much more verbose.

As I see it the major disadvantage with the current testAst output is that you can't see if an operator is unary right? And well it would probably help readability to space separate operators and operands.

how about if you could write this (z3 syntax):

    ASSERT_EQUALS("(return (new A))", testAst2("return new A;"));
    ASSERT_EQUALS("(return (+ (- 1) (/ 2 3)))", testAst2("return -1+2/3;"));

Or (updated rpn format):

    ASSERT_EQUALS("A new return", testAst2("return new A;"));
    ASSERT_EQUALS("1 u:- 2 3 / b:+ return", testAst2("return -1+2/3;"));

.. if `u:` is prepended it's a unary operator. if `b:` is prepended it's a binary operator. The `new` and `return` are unary operators however it seems we do not need `u:` for those.

I don't have a strong opinion but the z3 syntax is a bit better imho and we also use that in TestExprEngine.

Reply:
I'm all for using something that already exists if it does the job. When I was considering how to extend the simple ast text representation my concern was the need for escape sequences which I thought would make it hard to read the expression. For example in the z3 format you suggest above the parentheses indicate structure of the ast rather than tokens but in my test case ( is an operator. Effectively SimpleAst is the z3 syntax but with all tokens in "" (to show they're tokens, commas separate things instead of spaces, and with {} instead of (). so for example 
```(return (+ (- 1) (/ 2 3)))```
becomes
```{"return" {"+", {"-", "1"}, {"/", "2", "3"}}}```

This might be overkill though, how does z3 handle escaping?

Reply:
Oh just to answer the question asked, yes the main concern I have with the non-verbose ast string is the inability to distinguish binary and unary operators though I although think it's not ideal not being able to see token boundaries (which both your suggestions address).

Reply:
To clarify, Z3 does not have parentheses as operators. There are no function calls. And no control flow.

Do we need to escape parenthesis. Imagine:

    C++: dostuff(123);
    AST: (( dostuff 123)

Reply:
I believe that with minor tweaks the `testAst` can output "z3 syntax" instead.

Reply:
In the example ```(( dostuff 123)``` what identifies the second ```(``` as a token rather than being structural? Maybe I'm imagining an ambiguity that can't actually exist?

In any case the ast for ```"return new A();"``` would be ```(return (new (( A)))``` which isn't super easy to read but it's not impossible either. Maybe easier than the equivalent ```{"return", {"new", {"(", {"A"}}}}```

Reply:
> what identifies the second ( as a token rather than being structural

When it says "((", the first "(" is structural and the second "(" is the operator.

Reply:
well.. I think if it is easier to read we can replace "(" with a magic keyword such as "lpar"..